### PR TITLE
Use the error! macro for failure.

### PR DESCRIPTION
**Pull request by richard-uk1**
I love `quicli`, here is a suggestion on how to improve it (in the form of a PR).

Since we control the main method, there is no way that the logging init can fail. Given this, we can use the logging functionality for outputting any errors that are returned from the main method. This is what I have implemented in this PR.

Apart from this, there are a few small improvements, for example removing a warning in an edge case where all paths lead to an error, and adding tests for the different forms of the `main!` macro.

Let me know what you think!

### CODE REVIEW

File: src/main_macro.rs [hunk 1]
```diff
@@ -29,23 +31,68 @@
 ///     println!("{}", x);
 /// });
 /// ```
+///
+/// With command line arguments.
+///
+/// ```rust,ignore
```
Comment (killercup) [inline src/main_macro.rs]: why ignore and not no_run?

File: src/main_macro.rs [hunk 2]
```diff
@@ -29,23 +31,68 @@
...
+///
+/// main!(|args: Cli| {
+///     let x = read_file(&args.config)?;
+///     println!("{}", x);
+/// });
+/// ```
+///
+/// With command line arguments including a verbosity flag.
+///
+/// ```rust,ignore
```
Comment (killercup) [inline src/main_macro.rs]: why ignore and not no_run?

File: src/main_macro.rs [hunk 3]
```diff
@@ -29,23 +31,68 @@
...
+/// #[derive(Debug, StructOpt)]
+/// struct Cli {
+///     /// The config file for the cli app.
+///     #[structopt(long = "config", short = "c", parse(from_os_str))]
+///     config: PathBuf,
+///     #[structopt(flatten)]
+///     verbosity: Verbosity,
+/// }
+///
+/// main!(|args: Cli, log_level: Verbosity| {
```
Comment (killercup) [inline src/main_macro.rs]: This needs to be `verbosity` (lower case), as it refers to the field name, not the type. (I know this is awkward :))

File: src/main_macro.rs [hunk 4]
```diff
@@ -29,23 +31,68 @@
...
+/// ```
 #[macro_export]
 macro_rules! main {
     (|$args:ident: $cli:ty, log_level: $verbosity:ident| $body:expr) => {
         fn main() {
-            fn run() -> $crate::prelude::Result<()> {
-                let $args = <$cli>::from_args();
-                $args.$verbosity.setup_env_logger(&env!("CARGO_PKG_NAME"))?;
-
+            #[allow(unused_variables)]
```
Comment (killercup) [inline src/main_macro.rs]: Nice touch with the allows!

File: src/main_macro.rs [hunk 5]
```diff
@@ -29,23 +31,68 @@
...
-
+                #[allow(unreachable_code)]
                 Ok(())
             }
 
-            if let Err(e) = run() {
-                eprintln!("error: {}", e);
+            let $args = <$cli>::from_args();
+            // This cannot fail because we control the main method, so no-one else can register a
+            // log handler before us.
```
Comment (killercup) [inline src/main_macro.rs]: Is that the only reason log handlers can fail?

File: src/main_macro.rs [hunk 6]
```diff
@@ -29,23 +31,68 @@
...
+            $args.$verbosity.setup_env_logger(&env!("CARGO_PKG_NAME"))
+                .expect("logger setup should never fail, this is a bug in quicli");
```
Comment (killercup) [inline src/main_macro.rs]: Can you expand this expect a bit to be something like 

```rust
if let Err(e) = $args.$verbosity.setup_env_logger(&env!("CARGO_PKG_NAME")) {
    eprintln!("Failed to set up log handler. This is a bug in quicli!");
    eprintln!("Please report it: https://github.com/killercup/quicli/issues");
    eprintln!("Details: {}", e);
    ::std::process::exit(1);
}
```

(Same below)

**Review by killercup — CHANGES_REQUESTED** [review]
Thanks!

I'm not convinced that we want to use `error!` to output this, though. Since we use env_logger, all log lines will begin with a timestamp and the source module. Is that useful in this case?

We should expand the error message for failed logger setup, however. Because you are right, that would most likely be a bug in quicli – and the user can't do anything about it!

I'm not sure what your tests are testing? That these examples compile? That's what the examples in the `examples/` dir are for :) (they are part of the workspace and also contain tests with assertions that get run on CI)

File: src/main_macro.rs (hunk 2 above)
Comment (derekdreery) [reply src/main_macro.rs]: With no_run the tests fail. I think custom derive doesn't work in doctests.

File: src/main_macro.rs (hunk 5 above)
Comment (derekdreery) [reply src/main_macro.rs]: Yes - I looked through the methods and the only failure point is `log::set_boxed_logger` (called by `env_logger`), and that can only fail if a logger has already been set.

File: src/main_macro.rs (hunk 1 above)
Comment (derekdreery) [reply src/main_macro.rs]: Let's use the other comment to discuss this :)

File: src/main_macro.rs (hunk 6 above)
Comment (derekdreery) [reply src/main_macro.rs]: I dont' mind doing this, but it really should never fail, as the only way a logger init can fail is if there already is a logger installed, and we know that there isn't in this case.

I just included the `expect` as a minimal sanity check - I should probably just use `unwrap`.

**Comment by derekdreery** [comment]
Thanks for reviewing!

I hear what you're saying about logging, however I would argue that logging is still the correct way to return the error. `quicli`, and especailly `main!`, is very opinionated - that's kind of the point - and it's already blessed the use of logging in CLI apps. Maybe all log messages should not include the source? I think there is value in using a single strategy for messaging though - it's good for the user for all messages to follow the same format.

Having said all that if you still prefer, I'll happily change back to `eprintln` so you can merge the rest of the stuff in this PR.

About the tests - you're correct I didn't understand that the examples form a test suite so I'll remove those.
